File: C#/ProductExceptSelf.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        var semiSolution = 1;
        var zeroes = 0;
        foreach (var i in nums) {
            if (i != 0){
                semiSolution *= i;
            } else {
                zeroes++;
            }
        }
        int[] result = new int[nums.Length];
        for (int i = 0; i != nums.Length; ++i) {
            if (zeroes == 1) {
                if (nums[i] == 0) {
                    result[i] = semiSolution;
                } else {
                    result[i] = 0;
                }
            } else if (zeroes > 1) {
                result[i] = 0;
            } else {
                result[i] = semiSolution / nums[i];
            }
        }

        return result;
    }
}

File: Python/ProductExceptSelf.py
from functools import reduce

class Solution:
    def productExceptSelf(self, nums):
        length = len(nums)
        answer = [1] * length
        prefix = 1
        postfix = 1

        for i in range(length):
            answer[i] *= prefix
            prefix *= nums[i]

            answer[length - i - 1] *= postfix
            postfix *= nums[length - i - 1]

        return answer